src/Parser/DeathKnight/Unholy/Modules/Items/ColdHeartEfficiency.js

```diff
@@ -20,63 +30,106 @@ class ColdHeartEfficiency extends Analyzer {
-        }
-      }
-      if (this.buffColdHeart === 20) {
-        this.correctColdHeartCasts++;
-      }
-      //This is only here for double casting Cold Heart. If Cold Heart is casted again before it reaches 2 stacks, the event won't update.
-      this.buffColdHeart = 0;
-    }
+		this.totalColdHeartCasts++;
+	  
+		this.timeAtMaxStacks = event.timestamp - this.timeAtMaxStacksStart;
 
+		this.unholyStrengthRemaining = unholyStrengthDuration - (event.timestamp - this.unholyStrengthStart);
+		this.concordanceRemaining = concordanceDuration - (event.timestamp - this.concordanceStart);
+		this.khazgorothRemaining = khazgorothDuration - (event.timestamp - this.khazgorothStart);
+	  
+		//This checks wether or not any of the three buffs are about to fall off within the next 4 seconds.
+		if ((this.unholyStrengthRemaining > 0 && this.unholyStrengthRemaining < remainingDurationAllowed) || (this.concordanceRemaining > 0 && this.concordanceRemaining < remainingDurationAllowed) || (this.khazgorothRemaining > 0 && this.khazgorothRemaining < remainingDurationAllowed)){
+			if (this.buffColdHeart < coldHeartMaxStack && this.buffColdHeart >= coldHeartOptimalMinStack) {
+				this.correctColdHeartCasts++;
+			}
+		}else if (this.buffColdHeart < coldHeartMaxStack) {
+			this.castsTooEarly++;
+		}else if (this.buffColdHeart === coldHeartMaxStack && this.timeAtMaxStacks <= maxDurationAtMaxStacksAllowed) {
+			this.correctColdHeartCasts++;
+		}else if(this.buffColdHeart === coldHeartMaxStack){
+			this.castsTooLate++;
+		}
+		//This is only here for double casting Cold Heart. If Cold Heart is casted again before it reaches 2 stacks, the event won't update.
+		this.buffColdHeart = 0;
+    }
   }
-
   suggestions(when) {
     const castEfficiency = this.correctColdHeartCasts / this.totalColdHeartCasts;
     when(castEfficiency).isLessThan(0.8)
       .addSuggestion((suggest, actual, recommended) => {
-        return suggest(<span>You are casting <SpellLink id={SPELLS.CHAINS_OF_ICE.id} /> at non optimal times. You either want to cast <SpellLink id={SPELLS.CHAINS_OF_ICE.id} /> when at 20 stacks of <SpellLink id={SPELLS.COLD_HEART_BUFF.id} /> or when you are above 16 stacks and your buff <SpellLink id={SPELLS.UNHOLY_STRENGTH_BUFF.id} /> is about to run out.</span>)
+        return suggest(<wrapper> You are casting <SpellLink id={SPELLS.CHAINS_OF_ICE.id} /> at non optimal times. {this.castsTooLate} cast(s) were made too late and {this.castsTooEarly} cast(s) were made too early. You either want to cast <SpellLink id={SPELLS.CHAINS_OF_ICE.id} /> when at 20 stacks of <SpellLink id={SPELLS.COLD_HEART_BUFF.id} /> ASAP, or when you are above 13 stacks and any of your buffs <SpellLink id={SPELLS.UNHOLY_STRENGTH_BUFF.id}/> or <SpellLink id={SPELLS.CONCORDANCE_OF_THE_LEGIONFALL_STRENGTH.id} />  or <SpellLink id={SPELLS.KHAZGOROTHS_SHAPING.id} /> are about to run out. You also don't want to hold <SpellLink id={SPELLS.CHAINS_OF_ICE.id} /> at 20 stacks for too long.</wrapper>)
```

Comment:
It's `<Wrapper>`, and you'll need to `import Wrapper from 'common/Wrapper';` at the top... as written I'm pretty sure this will crash.